sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,343 @@
+
+                ParameterData paramData;
+                if (!parametersToCheck.TryGetValue(key, out paramData) ||
+                    !paramData.ShouldReportOn)
+                {
+                    continue;
+                }
+
+                if (idName.Parent is EqualsValueClauseSyntax ||
+                    idName.Parent is AssignmentExpressionSyntax)
+                {
+                    paramData.ShouldReportOn = false;
+                    continue;
+                }
+
+                var symbolUsedAs = FindParameterUseAsType(idName, context.SemanticModel);
+                if (symbolUsedAs != null)
+                {
+                    paramData.AddUsage(symbolUsedAs);
+                }
+            }
+
+            return parametersToCheck
+                .Values
+                .Select(p => p.GetRuleViolation())
+                .WhereNotNull();
+        }
+
+        private static bool IsTrackedParameter(IParameterSymbol parameterSymbol)
+        {
+            var type = parameterSymbol.Type;
+            var isExcluded =
+                type.DerivesFrom(KnownType.System_Array) ||
+                type.IsValueType ||
+                type.Is(KnownType.System_String);
+
+            return !isExcluded;
+        }
+
+        private ITypeSymbol FindParameterUseAsType(IdentifierNameSyntax idRef, SemanticModel semanticModel)
+        {
+            if (idRef.Parent is ConditionalAccessExpressionSyntax)
+            {
+                var cond = idRef.Parent as ConditionalAccessExpressionSyntax;
+                var binding = cond.WhenNotNull as MemberBindingExpressionSyntax;
+
+                if (binding != null)
+                {
+                    var name = binding.Name;
+                    var accessedMember = semanticModel.GetSymbolInfo(name).Symbol;
+
+                    var propertySymbol = accessedMember as IPropertySymbol;
+                    if (propertySymbol != null)
+                    {
+                        return HandleProperty(idRef, propertySymbol);
+                    }
+
+                    return FindOriginatingSymbol(accessedMember);
+                }
+                else if (cond.WhenNotNull is InvocationExpressionSyntax)
+                {
+                    var inv = (cond.WhenNotNull as InvocationExpressionSyntax).Expression as MemberBindingExpressionSyntax;
+
+                    if (inv != null)
+                    {
+                        var invocationSymbol = semanticModel.GetSymbolInfo(inv).Symbol;
+                        return HandleInvocation(idRef, invocationSymbol, semanticModel);
+                    }
+                }
+            }
+            else if (idRef.Parent is MemberAccessExpressionSyntax)
+            {
+                var invocation = idRef.Parent.Parent as InvocationExpressionSyntax;
+                if (invocation != null)
+                {
+                    var invocationSymbol = semanticModel.GetSymbolInfo(idRef.Parent.Parent).Symbol;
```

Comment:
`invocation` is `idRef.Parent.Parent` and when you are in `invocation != null` you can safely use `semanticModel.GetSymbolInfo(invocation)` instead of again `idRef.Parent.Parent`